File: 성경민/Week2/BOJ_10845_큐.java
/*
10845(큐)
push X: 정수 X를 큐에 넣는 연산이다.
pop: 큐에서 가장 앞에 있는 정수를 빼고, 그 수를 출력한다. 만약 큐에 들어있는 정수가 없는 경우에는 -1을 출력한다.
size: 큐에 들어있는 정수의 개수를 출력한다.
empty: 큐가 비어있으면 1, 아니면 0을 출력한다.
front: 큐의 가장 앞에 있는 정수를 출력한다. 만약 큐에 들어있는 정수가 없는 경우에는 -1을 출력한다.
back: 큐의 가장 뒤에 있는 정수를 출력한다. 만약 큐에 들어있는 정수가 없는 경우에는 -1을 출력한다.

15
push 1
push 2
front
back
size
empty
pop
pop
pop
size
empty
pop
push 3
empty
front
*/

import java.util.*;
import java.io.*;
public class Main {

    public static void main(String[] args) throws NumberFormatException, IOException {
        //Queue 생성
        Queue<Integer> q = new LinkedList<>();

        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));

        int x = Integer.parseInt(br.readLine());

        //Queue에서 인덱스 쓸 수 없으므로 마지막 칸에 들어가는 숫자를 last 변수에 넣음
        int last = 0;

        for(int i=0; i<x; i++) {
            String[] str = br.readLine().split(" ");

            switch(str[0]) {
                case ("push") :
                    q.add(Integer.parseInt(str[1]));
                    last = Integer.parseInt(str[1]);
                    break;

                case ("pop") :
                    if(q.isEmpty()) {
                        bw.write("-1" + "\n");
                    } else {
                        bw.write(q.poll() + "\n");
                    }
                    break;

                case ("size") :
                    bw.write(q.size() + "\n");
                    break;

                case ("empty") :
                    if(q.isEmpty()) {
                        bw.write("1" + "\n");
                    } else {
                        bw.write("0" + "\n");
                    }
                    break;

                case ("front") :
                    if(q.isEmpty()) {
                        bw.write("-1" + "\n");
                    } else {
                        bw.write(q.peek() + "\n");
                    }
                    break;

                case ("back") :
                    if(q.isEmpty()) {
                        bw.write("-1" + "\n");
                    } else {
                        bw.write(last + "\n");
                    }
                    break;
            }
        }
        System.out.println(q);
        bw.flush();
        bw.close();
    }
}

File: 오주영/Week2/BOJ_10845_큐.py
import sys  #입력을 stdin 형태로 받기 위해 불러온다.


S = []

#queue 기능 구현 ~ list를 이용해서 
def push(S : list, any):    
    # d = int(any)     #생각해보니 int를 쓸 이유가?
    S.insert(0,any)    #제일 왼쪽에 넣는다 ; 이러면 stack의 뼈대를 재활용할 수 있다.     

def pop(S):
    if S == []:
        print(-1)
    else:
        value = S[len(S)-1] #S의 가장 오른쪽 값이다 
        S.pop()             #S에서 가장 오른쪽 값을 지운다
        print(value)

def front(S):
    if S == []:
        print(-1)
    else:
        print(S[-1])        #음수 인덱스는 -1부터 오른쪽에서 왼쪽

def back(S):
    if S == []:
        print(-1)
    else:
        print(S[0])

def empty(S):
    if S==[]:
        print(1)
    else:
        print(0)

def size(S):
    print(len(S))




N = int(sys.stdin.readline()) #문자열을 int로 형변환 하면 개행문자 지워짐

for i in range(N):
    f = sys.stdin.readline().split() #문자열을 공백단위로 / f[0]이 시동어가 된다

    if f[0]=='push':
        push(S,f[1])
    elif f[0]=='pop':
        pop(S)
    elif f[0]=='front':
        front(S)
    elif f[0]=='back':
        back(S)
    elif f[0]=='empty':
        empty(S)
    elif f[0]=='size':
        size(S)
    
    '''함수명과 f[0]가 같은데 코드를 더 짧게 하는 방법이 있을까?'''
